# TEIRI 

A developed pipeline to identify TE-initiated RNAs by integrating short-read RNA-seq, long-read RNA-seq, CAGE, and RAMPAGE datasets.

This repository contains the pipeline (script/) and TE-initiated RNA annotaion files (Annotation/).

## Software requirements

Python 3.x.x, BEDTools, gffread

## Usage

### Input files 

```
GTF files: 
    GTF files which could be generated by stringtie from short-read RNA-seq.
BED12 file: 
    the BED12 file which was generated by flair correct. In brief, the reads of long-read RNA-seq were aligned to the genome using minmap2, and the sorted bam was converted to the BED12. 
    Then, misaligned splice sites were corrected using genome annotations and short-read splice junctions by flair correct.
TSS signal: 
    the counts of unique mapping reads (CAGE and/or RAMPAGE) supporting the tss in tsv format (at least 5 fields: chrom, chromStart, chromEnd, strand, and score) 
TE reference bed: 
    TE annotation in BED format (at least 6 fields: chrom, chromStart, chromEnd, ID, name, and strand).
Reference GTF: 
    reference GTF file for gene annotations.
```

### TEIRI_correct.py

TEIRI_correct.py extracts and filters the first exon of transcript. TSSs of the first exons are then corrected based on the counts of reads supporting the TSS (including long-read RNA-seq, CAGE, and RAMPAGE).

```
usage: TEIRI_correct.py [-h] [-i GTF_LIST] [-r REFERENCE_GTF] [-l CORRECTED_BED12] [-c TSS_SCORE] [--TE_anno TE_ANNO] [--max_exon1_length MAX_EXON1_LENGTH] [--min_exon1_length MIN_EXON1_LENGTH]
                        [--tss_window TSS_WINDOW] [--max_tss MAX_TSS] [--TGS_weight TGS_WEIGHT] [--min_NGS_ratio MIN_NGS_RATIO] [--min_TGS_reads MIN_TGS_READS] [--threshold THRESHOLD]
                        [--min_NGS_ratio_SE MIN_NGS_RATIO_SE] [--min_TGS_reads_SE MIN_TGS_READS_SE] [--SE_threshold SE_THRESHOLD] [-p PREFIX]

TEIRI_correct

options:
  -h, --help            show this help message and exit
  -i GTF_LIST, --gtf_list GTF_LIST
                        A text file with a list of GTF files (required)
  -r REFERENCE_GTF, --reference_gtf REFERENCE_GTF
                        Reference genome annotation in GTF format (required)
  -l CORRECTED_BED12, --corrected_bed12 CORRECTED_BED12
                        The flair-corrected bed12 file
  -c TSS_SCORE, --tss_score TSS_SCORE
                        A tsv file with the counts of unique mapping reads supporting the tss (CAGE and/or RAMPAGE)
  --TE_anno TE_ANNO     TE annotation in BED format (required)
  --max_exon1_length MAX_EXON1_LENGTH
                        The max length of first exon (default: 2588)
  --min_exon1_length MIN_EXON1_LENGTH
                        The min length of first exon (default: 20)
  --tss_window TSS_WINDOW
                        The window size used to calculate the weight score (default: 50)
  --max_tss MAX_TSS     Maximum number of TSS picked per first exon (default: 2)
  --TGS_weight TGS_WEIGHT
                        The weight of TGS reads supporting the TSS (default: 1)
  --min_NGS_ratio MIN_NGS_RATIO
                        The min NGS ratio supporting the first exon (default: 0.05)
  --min_TGS_reads MIN_TGS_READS
                        The min TGS reads supporting the first exon (default: 2)
  --threshold THRESHOLD
                        The threshold of weight score for TSS (default: 10)
  --min_NGS_ratio_SE MIN_NGS_RATIO_SE
                        The min NGS ratio supporting the single exon (default: 0.25)
  --min_TGS_reads_SE MIN_TGS_READS_SE
                        The min TGS reads supporting the single exon (default: 10)
  --SE_threshold SE_THRESHOLD
                        The threshold of weight score for TSS of the single exon (default: 20)
  -p PREFIX, --prefix PREFIX
                        Prefix for output file (default: TEIRI)
```

### TEIRI_merge.py

TEIRI_merge.py is designed for transcript merge using corrected TSSs, which is suited for multiple samples in one condition.

```
usage: TEIRI_merge.py [-h] [-i GTF_LIST] [-r REFERENCE_GTF] [-l CORRECTED_BED12] [--corrected_tss CORRECTED_TSS] [--corrected_tss_single CORRECTED_TSS_SINGLE] [--TGS_weight TGS_WEIGHT]
                      [--illumina_threshold ILLUMINA_THRESHOLD] [--nanopore_threshold NANOPORE_THRESHOLD] [--ref_transcript_length REF_TRANSCRIPT_LENGTH] [--max_transcripts MAX_TRANSCRIPTS]
                      [--trunctated_exclude TRUNCTATED_EXCLUDE] [--min_transcript_length MIN_TRANSCRIPT_LENGTH] [-p PREFIX]

TEIRI_merge

options:
  -h, --help            show this help message and exit
  -i GTF_LIST, --gtf_list GTF_LIST
                        A text file with a list of GTF files (required)
  -r REFERENCE_GTF, --reference_gtf REFERENCE_GTF
                        Reference genome annotation in GTF format (required)
  -l CORRECTED_BED12, --corrected_bed12 CORRECTED_BED12
                        The flair-corrected bed12 file
  --corrected_tss CORRECTED_TSS
                        A tsv file with the corrected TE-derived TSSs (TEIRI_correct.py generated)
  --corrected_tss_single CORRECTED_TSS_SINGLE
                        A tsv file with the corrected TE-derived TSSs for the single-exon transcript (TEIRI_correct.py generated)
  --TGS_weight TGS_WEIGHT
                        The weight of TGS reads supporting the transcript (default: 1000)
  --illumina_threshold ILLUMINA_THRESHOLD
                        The min NGS ratio supporting the transcript (default: 0.05)
  --nanopore_threshold NANOPORE_THRESHOLD
                        The min TGS reads supporting the transcript (default: 2)
  --ref_transcript_length REF_TRANSCRIPT_LENGTH
                        The average length of reference transcripts (default: 2725)
  --max_transcripts MAX_TRANSCRIPTS
                        The max counts of transcripts for a TE-initiated RNA (default: 10)
  --trunctated_exclude TRUNCTATED_EXCLUDE
                        Truncated transcripts were excluded, as they may represent fragments of the full-length transcript (default: True)
  --min_transcript_length MIN_TRANSCRIPT_LENGTH
                        The min transcript length (default: 200)
  -p PREFIX, --prefix PREFIX
                        Prefix for output file (default: TEIRI)
```

### TEIRI_consolidate.py

TEIRI_consolidate.py is designed for the consolidation of TE-initiated RNAs in multiple conditions, such as different tissues.

```
usage: TEIRI_consolidate.py [-h] [-i GTF_LIST] [-r REFERENCE_GTF] [--TE_anno TE_ANNO] [--tss_merge_distance TSS_MERGE_DISTANCE] [--min_exon_length MIN_EXON_LENGTH] [-p PREFIX]

TEIRI_consolidate

options:
  -h, --help            show this help message and exit
  -i GTF_LIST, --gtf_list GTF_LIST
                        A text file with a list of GTF files (required)
  -r REFERENCE_GTF, --reference_gtf REFERENCE_GTF
                        Reference genome annotation in GTF format (required)
  --TE_anno TE_ANNO     TE annotation in BED format (required)
  --tss_merge_distance TSS_MERGE_DISTANCE
                        The min distance to merge TE-derived TSSs of different GTF files (default: 50)
  --min_exon_length MIN_EXON_LENGTH
                        The min length of first exon (default: 20)
  -p PREFIX, --prefix PREFIX
                        Prefix for output file (default: TEIRI)
```
